File: src/Wc3Handle/Timer.ts
/** @noSelfInFile */

import { Action, ActionList, Log } from "../Wc3Utils/index";
import { Handle } from "./Handle";

export class Timer extends Handle<jtimer> {
    constructor(){super(CreateTimer())}

    public static get(id: jtimer | number){
        let instance = Handle.get(id)
        if (!instance){return}
        if (Handle.getWc3Type(instance.handle) != 'timer'){
            Log.err('Timer: got wrong type of handle.', 2)
        }
        return instance as Timer
    }
    public static getExpired(){return Timer.get(GetExpiredTimer())}

    public get timeout(){return this._timeout}
    public get periodic(){return this._periodic}

    public start(timeout: number, periodic:boolean){
        this._timeout = timeout
        this._periodic = periodic
        TimerStart(this.handle, timeout, periodic, Timer.runActions)
    }
    public pause(){PauseTimer(this.handle)}
    public resume(){ResumeTimer(this.handle)}
    public destroy(){
        PauseTimer(this.handle)
        DestroyTimer(this.handle)
        super.destroy()
    }

    public addAction(callback: (timer: Timer)=>void){
        return this._actions.add(callback)
    }
    public removeAction(action: Action<[Timer], void> | undefined){
        this._actions.remove(action)
    }
    private static runActions(this: void){
        let timer = Timer.getExpired()
        timer?._actions.run(timer)
    }

    private _timeout = -1;
    private _periodic = false;
    private _actions = new ActionList<[Timer]>();
}

File: src/Wc3HandleLua/Handle/Timer.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local Action = Wc3Utils.Action or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')

---@type HandleClass
local Handle = require('Handle') or error('')

--=======
-- Class
--=======

local Timer = Class.new('Timer', Handle)
---@class Timer : Handle
local public = Timer.public
---@class TimerClass : HandleClass
local static = Timer.static
---@type TimerClass
local override = Timer.override
local private = {}

--=========
-- Static
--=========

---@param child Timer | nil
---@return Timer
function override.new(child)
    if child then isTypeErr(child, Timer, 'child') end

    local instance = child or Class.allocate(Timer)
    instance = Handle.new(CreateTimer(), private.destroyTimerHandle, instance)

    return instance
end

--========
-- Public
--========

---@param timeout number
---@param periodic boolean
---@param callback Callback
function public:start(timeout, periodic, callback)
    isTypeErr(self, Timer, 'self')
    isTypeErr(timeout, 'number', 'timeout')
    isTypeErr(periodic, 'boolean', 'periodic')
    isTypeErr(callback, 'function', 'callback')

    local action = Action.new(callback, self)
    TimerStart(self:getData(), timeout, periodic, function() action:run() end)
end

function public:pause()
    isTypeErr(self, Timer, 'self')
    PauseTimer(self:getData())
end

function public:resume()
    isTypeErr(self, Timer, 'self')
    ResumeTimer(self:getData())
end

--=========
-- Private
--=========

function private.destroyTimerHandle(handle)
    PauseTimer(handle)
    DestroyTimer(handle)
end

return static